### PR TITLE
sql: Refactor SINK parsing

### LINKED ISSUE #14206: sql: CREATE SINK parsing needs significant refactor
### What version of Materialize are you using?

v0.26.1-dev (8b9c797ea)

### How did you install Materialize?

Built from source

### What is the issue?

This is valid syntax for a `CREATE SINK` statement:

```sql
CREATE SINK foo FROM bar INTO KAFKA BROKER 'baz' TOPIC 'topic' WITH
```

This should be refactored before we make sinks broadly accessible again.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by sploiselle**
`SINK` parsing had a few rough corners.

### Motivation

This PR adds a known-desirable feature. Closes #14206

### Tips for reviewer

I am going to squash this change, so apologies for the slight mess at the tail end of the commit history.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - `CREATE SINK`'s `SNAPSHOT` option now behaves as other `WITH` options and requires its values to be wrapped in parentheses.
  - To disable `SNAPSHOT` when creating a `SINK` use `CREATE SINK...WITH (SNAPSHOT = false)`.
  - `CONFLUENT SCHEMA REGISTRY CONNECTION` references now support options, such as `AVRO KEY FULLNAME`.
  - `KAFKA CONNECTION` options (e.g. `RETENTION BYTES`) can no longer be prefixed by `WITH`. For example, users now type `CREATE SOURCE...KAFKA CONNECTION (RETENTION BYTES = 1000)...`.

### CODE REVIEW

File: src/sql-parser/src/parser.rs [hunk 1]
```diff
@@ -2328,30 +2366,40 @@ impl<'a> Parser<'a> {
...
-            with_options,
             format,
             envelope,
-            with_snapshot,
             if_not_exists,
+            with_options,
         }))
     }
 
+    fn parse_create_sink_options(&mut self) -> Result<CreateSinkOption<Raw>, ParserError> {
```
Comment (bkirwi) [inline src/sql-parser/src/parser.rs:2390]: nit:

```suggestion
    fn parse_create_sink_option(&mut self) -> Result<CreateSinkOption<Raw>, ParserError> {
```

File: src/sql-parser/tests/testdata/ddl [hunk 2]
```diff
@@ -575,39 +575,39 @@ CREATE SOURCE psychic FROM POSTGRES CONNECTION pgconn PUBLICATION 'red'
 CreateSource(CreateSourceStatement { name: UnresolvedObjectName([Ident("psychic")]), col_names: [], connection: Postgres { connection: Name(UnresolvedObjectName([Ident("pgconn")])), publication: "red", details: None }, legacy_with_options: [], include_metadata: [], format: None, envelope: None, if_not_exists: false, key_constraint: None, with_options: [] })
 
 parse-statement
-CREATE SINK foo FROM bar INTO KAFKA CONNECTION baz TOPIC 'topic' WITH (replication_factor = 7, retention_ms = 10000, retention_bytes = 10000000000) FORMAT BYTES
+CREATE SINK foo FROM bar INTO KAFKA CONNECTION baz (REPLICATION FACTOR = 7, RETENTION MS = 10000, RETENTION BYTES = 10000000000) TOPIC 'topic' FORMAT BYTES
```
Comment (bkirwi) [inline src/sql-parser/tests/testdata/ddl:578]: FWIW, it reads a bit weirdly to me to have the topic configs follow the connection object and not the topic name. I understand the old `WITH` was ambiguous and has to go, but would it be possible to stick with the old order? (A la `CREATE SINK foo FROM bar INTO KAFKA CONNECTION baz TOPIC 'topic' (REPLICATION FACTOR = 7, RETENTION MS = 10000, RETENTION BYTES = 10000000000) FORMAT BYTES`)

(Non-blocking, ofc... I understand this is personal taste.)

**Review by bkirwi — COMMENTED** [review]
Heads up that this will conflict rather violently with #14526. Not much to be done about that, though.

File: src/sql-parser/tests/testdata/ddl (hunk 2 above)
Comment (sploiselle) [reply src/sql-parser/tests/testdata/ddl:578]: I have no opinion on this whatsoever; if you can get anyone else to agree this is a good idea, glad to make the change. @brennanvincent you've worked with Kafka a lot--wdyt?

Comment (brennanvincent) [reply src/sql-parser/tests/testdata/ddl:578]: Yeah, putting the topic first makes sense to me

Comment (sploiselle) [reply src/sql-parser/tests/testdata/ddl:578]: Thanks for the suggestion, @bkirwi! Will do this before merging, but am going to wait and see if there's any other feedback.

Comment (benesch) [reply src/sql-parser/tests/testdata/ddl:578]: I'm also pro doing this! I think of the `CREATE SOURCE` and `CREATE SINK` statements as having three stanzas:

  * Connection
  * Format
  * Envelope
  * General options (`WITH`)

Where each stanza is followed by enum-style options in parentheticals `()`.

In a perfect world I think the Kafka connection's `TOPIC` field and the PostgreSQL connection's `PUBLICATION` field would be plain ol' enum options in the parentheticals, but we've probably caused enough `CREATE SOURCE`/`CREATE SINK` chaos for a lifetime.

Comment (sploiselle) [reply src/sql-parser/tests/testdata/ddl:578]: Moving topic (and publication) into the options isn't hard and I'd be glad to do it while the wounds are fresh :D

Comment (bkirwi) [reply src/sql-parser/tests/testdata/ddl:578]: `TOPIC` is a bit special in that it's always required, whereas i think everything else has a sensible default? but i hear the argument for sure.

**Review by benesch — APPROVED** [review]
So good! Thank you, @sploiselle!

File: src/sql-parser/tests/testdata/ddl (hunk 2 above)
Comment (benesch) [reply src/sql-parser/tests/testdata/ddl:578]: Yup yup, but we have a bunch of other places where we have mandatory enum style options. E.g., `USER` for PostgreSQL connections. It’s actually really nice when you have multiple mandatory options to be able to specify them in any order; in the current syntax, you have to remember the exact right order.

**Pull request merged**